src/firstpass.rs

```diff
@@ -1183,6 +1217,120 @@ impl<'a, 'b> FirstPass<'a, 'b> {
+            // Characters to be excluded:
+            //  * `{` and `}`: special characters to open and close an attribute block.
+            //  * `\\`: a special character to escape many characters and disable some syntaxes.
+            //      + Handling of this escape character differs among markdown processors.
+            //      + Escaped characters will be separate text node from neighbors, so
+            //        it is not easy to handle unescaped string and trim the trailing block.
+            //  * `<` and `>`: special characters to start and end HTML tag.
+            //      + No known processors converts `{#<i>foo</i>}` into
+            //        `id="&lt;i&gt;foo&lt;/&gt;"` as of this writing, so hopefully
+            //        this restriction won't cause compatibility issues.
+            //  * `\n` and `\r`: a newline character.
+            //      + Setext heading can have multiple lines. However it is hard to support
+            //        attribute blocks that have newline inside, since the parsing proceeds line by
+            //        line and lines will be separate nodes even they are logically a single text.
+            !matches!(b, b'{' | b'}' | b'<' | b'>' | b'\\' | b'\n' | b'\r')
+        });
+        if ix == begin_ix {
+            // `{` is not found. No attribute blocks available.
+            return (end_ix, None);
+        }
+        let attr_block_open = ix - 1;
+        if bytes[attr_block_open] != b'{' {
+            // `{` is not found. No attribute blocks available.
+            return (end_ix, None);
+        }
+
+        (
+            attr_block_open,
+            Some(&self.text[attr_block_open..=attr_block_close]),
+        )
+    }
+
+    /// Extracts and parses a heading attribute block if exists.
+    ///
+    /// Returns `(end_offset_of_heading_content, (id, classes))`.
+    ///
+    /// If `header_end` is less than or equal to `header_start`, the given
+    /// input is considered as empty.
+    fn extract_and_parse_heading_attribute_block(
+        &mut self,
+        header_start: usize,
+        header_end: usize,
+    ) -> (usize, (Option<&'a str>, Vec<&'a str>)) {
+        if header_start >= header_end {
+            return (header_end, (None, Vec::new()));
+        }
+
+        let bytes = self.text.as_bytes();
+
+        // extract the trailing attribute block
+        let (content_end, attr_block) = if self.options.contains(Options::ENABLE_HEADING_ATTRIBUTES)
+        {
+            let (mut content_end, attr_block) =
+                self.extract_attribute_block_from_header_text(header_start, header_end);
+            content_end -=
+                scan_rev_while(&bytes[header_start..content_end], is_ascii_whitespace_no_nl);
+            (content_end, attr_block)
+        } else {
+            (header_end, None)
+        };
+
+        // parse inside the attribute block
+        let id_classes = if let Some(attr_block) = attr_block {
+            debug_assert!(
```

Comment:
Does this mean that the slice returned by `extract_attribute_block_from_header_text` is supposed to include the curly braces? If we changed that function to not do that, we don't need this check and don't need to take a subslice here at all :-)